### ISSUE TITLE
perf: PSV decode (PackedSfenValue::decode → ShogiBoard) の SIMD 化 (D4)

### ISSUE THREAD

**Issue by SH11235** [SH11235/tatara#199]
## 背景

`crates/shogi-format/src/packed_sfen.rs` `PackedSfenValue::decode()` は 40 byte の packed
sfen を 81 square の `ShogiBoard` に展開する。bit-packed の `huffman` style encoding
(king pos 7 bit + 持駒 + 盤上駒 variable-length) を 1 PSV につき逐次 decode。

worker hot path で `Batch::push_decoded` の直前に呼ばれる (`scratch[psv]` → `psv.decode()`
→ board)。`batch=65536` × `~85 batch/sec` = ~890K decode/sec、CPU dataloader 工程の主要
一つ。

## 提案

PSV record decode は per-record 独立 → batch ごと並列 + SIMD で multi-record 同時 decode。

- 4-8 record を SIMD lane に並べる pattern: AVX2 (256bit、4 × i64) で 4 record 同時 unpack
- variable-length encoding なので lane 間で進行率が違う → mask + serial fallback 必要
- 実装複雑、bit-twiddling 多い

代替案: `decode()` の中の scalar bit shift / lookup を最適化 (SIMD 化せずとも +20-30%
gain あるかも、`perf record` の hot function 確認後判断)。

## 想定 gain

PSV decode が CPU dataloader 時間の 30-40% (要計測) と想定、SIMD 4x で +10-15%。
ROI は #198 (sparse 抽出 SIMD) より低い見込み (decode より sparse 抽出の方が op 数多いため)。

## NO-GO 基準

- 着手前 `perf record` で PSV decode が CPU dataloader 時間の 20% 未満なら NO-GO
- 実装後 +3% 未満なら revert
- bit-identical 保証必須 (decode 結果が scalar と完全一致、board state byte-by-byte 同じ)

## 受け入れ基準

- [ ] `bash scripts/local-ci.sh` PASS
- [ ] `crates/shogi-format` の既存 PSV decode test 全 PASS (`tests/psv_smoke.rs` 等)
- [ ] Simple smoke 全 4 段 PASS、量子化 / raw ckpt bit-identical
- [ ] `256x2-32-32 halfkp` で +3% 以上 pos/s 改善
- [ ] LayerStack regression なし
- [ ] codex:codex-rescue 局所 review APPROVED

## 関連

- 前 Issue: #188 D2-D7 deferred、#194-#198 (本 wave の他 D / N 候補)、#142 (PSV → GPU decode
  本格 refactor、本 SIMD で十分なら #142 不要、scope 縮小)
- 参考: bullet-shogi の Rust packed sfen decode 実装、nnue-pytorch C++ decode

**Comment by SH11235** [comment]
## 2026-05-24 更新: SIMD 4x → Fast-Huffman 6-bit peek LUT へ pivot

board 1-pass refactor (#244 / commit 8608aba) 後の WSL ext4 (RTX 2070 SUPER + i9-10900X、PSV warm pagecache) で `samply` 計測した結果を反映。詳細は `docs/experiments/2026-05-24-wsl-ext4-reference-and-decode-profile.md` の section 3。

### Profile 根拠

`256x2-32-32 halfkp THREADS=20` で 20 worker 集計 sample = 161,581 (worker busy 31%):

| 領域 | 累積 self % | 主要 hot lines |
|---|---:|---|
| `from_packed_sfen` 系 | **38.5%** | `packed_sfen.rs:79/85` (BitStream::read_bit ~6.7%)、`:441` (Huffman match)、decode_board_piece (~5%)、board store (~8%) |
| `map_features_board` | 31.8% | refactor 済 |

→ NO-GO 基準「PSV decode が CPU 時間の 20% 未満」を満たさず、algorithmic 改善余地が大きいことを確認。

### 算法 pivot

元 Issue の SIMD 4-record lane 並列ではなく、**1 record 内の Huffman scan 自体を 8-entry 線形 → 6-bit peek lookup (64 entry table)** で削減する形に変える。理由:

- `BitStream::read_bit` を 1 piece あたり 1-8 回呼び、各 read_bit ごとに HUFFMAN_TABLE 8 entry を線形 scan = decode の支配的 work
- 6 bit (board 最大 Huffman 長) を 1 回 peek → LUT で `(piece_idx, huffman_len)` を 1 lookup で取得 → cursor を huffman_len 進める → 残り promote / color フラグ bit を読む
- hand decode は 5 bit peek + 32-entry LUT で同パターン
- 1 piece あたり 6-8 回 read_bit + 8 回 linear scan が **1 回 peek + 1 lookup + 1-2 回 read_bit** に削減 → 概算 5-6x の per-piece 高速化見込み

SIMD なしで実装が単純 (LUT は const fn で compile 時生成、bit-identical 検証が table-driven test で網羅可能)、と #244 と同じ pattern。

### 受け入れ基準 (WSL ext4 baseline、3080 Ti は ssh 占有解除後)

- [ ] `bash scripts/local-ci.sh` PASS (`CUDA_OXIDE_TARGET=sm_75` export 済)
- [ ] `crates/shogi-format` の既存 PSV decode test 全 PASS + 全 Huffman piece × promote × color の table-driven test 追加
- [ ] Simple smoke 4 段 PASS、量子化 `.bin` が main vs branch で byte-identical (`cmp`)。`.ckpt` は GPU 非決定性で比較しない
- [ ] WSL ext4 (`256x2-32-32 halfkp THREADS=10`、`bench-pos.env` の ext4 PSV) で baseline 2.78M pos/s に対して +3% 以上 (本機は GPU saturate 状態なので end-to-end gain が出ない可能性あり、その場合は samply で per-thread decode 時間の short化を別途確認)
- [ ] codex:codex-rescue + Claude general-purpose 局所 review 両 APPROVED
- [ ] 3080 Ti 計測は ssh 占有解除後 PR comment で追記 (本 PR の merge ブロッカーにはしない)

### 再現コマンド

```bash
export CUDA_OXIDE_TARGET=sm_75
bash scripts/local-ci.sh

# unit test (Huffman decode bit-identical):
cargo test --release -p shogi-format

# smoke 4 段:
target/release/nnue-train smoke-simple

# WSL ext4 bench (baseline 2.78M、本機 RTX 2070 SUPER + i9-10900X):
THREADS=10 ARCH=simple SIMPLE_ARCH=256x2-32-32 \
  scripts/bench-pos.sh   # DATA は scripts/bench-pos.env で指定済 (ext4)
```

### 関連

- profile doc: `docs/experiments/2026-05-24-wsl-ext4-reference-and-decode-profile.md`
- prior refactor (同パターン): #244 / commit 8608aba (board 1-pass)
